rule/string-regex.go

```diff
@@ -0,0 +1,233 @@
+		})
+	}
+}
+
+func (w lintStringRegexRule) parseArgument(argument interface{}, ruleNum int) (scope stringRegexSubruleScope, regex *regexp.Regexp, errorMessage string) {
+	g, ok := argument.([]interface{}) // Cast to generic slice first
+	if !ok {
+		panic(fmt.Sprintf("unable to parse argument %d", ruleNum))
+	}
+	var rule []string
+	for i, obj := range g {
+		val, ok := obj.(string)
+		if !ok {
+			panic(fmt.Sprintf("unable to parse argument %d, option %d", ruleNum, i))
+		}
+		rule = append(rule, val)
+	}
+
+	// Parse rule scope
+	scope = stringRegexSubruleScope{}
+	matches := parseStringRegexScope.FindStringSubmatch(rule[0])
+	if matches == nil {
+		panic(fmt.Sprintf("unable to parse rule scope (argument %d, option 0)", ruleNum))
+	}
+	scope.Func = matches[1]
+	if len(matches[2]) > 0 {
+		var err error
+		scope.Argument, err = strconv.Atoi(matches[2])
+		if err != nil {
+			fmt.Println(len(matches))
+			panic(fmt.Sprintf("unable to parse rule scope argument number (argument %d, option 0)", ruleNum))
+		}
+	}
+	if len(matches[3]) > 0 {
+		scope.Field = matches[3]
+	}
+
+	// Strip / characters from the beginning and end of rule[1] before compiling
+	regex, err := regexp.Compile(rule[1][1 : len(rule[1])-1])
+	if err != nil {
+		panic(fmt.Sprintf("unable to compile %s as regexp (argument %d, option 1)", rule[1], ruleNum))
+	}
+
+	// Use custom error message if provided
+	if len(rule) == 3 {
+		errorMessage = rule[2]
+	}
+	return scope, regex, errorMessage
+}
+
+// #endregion
+
+// #region Node traversal
+
+func (w lintStringRegexRule) Visit(node ast.Node) ast.Visitor {
+	// First, check if node is a call expression
+	call, ok := node.(*ast.CallExpr)
+	if !ok {
+		return w
+	}
+
+	// Get the name of the call expression to check against rule scope
+	callName, ok := w.getCallName(call)
+	if !ok {
+		return w
+	}
+
+	for _, rule := range w.rules {
+		if rule.Scope.Func == callName {
+			rule.Apply(call)
+		}
+	}
+
+	return w
+}
+
+// Return the name of a call expression in the form of package.Func or Func
+func (w lintStringRegexRule) getCallName(call *ast.CallExpr) (callName string, ok bool) {
+	if ident, ok := call.Fun.(*ast.Ident); ok {
+		// Local function call
+		return ident.Name, true
+	} else if selector, ok := call.Fun.(*ast.SelectorExpr); ok {
+		// Scoped function call
+		scope, ok := selector.X.(*ast.Ident)
+		if !ok {
+			return "", false
+		}
+		return scope.Name + "." + selector.Sel.Name, true
+	} else {
+		return "", false
+	}
+}
+
+// #endregion
+
+// #region Linting logic
+
+// Apply a single regex rule to a call expression (should be done after verifying the that the call expression matches the rule's scope)
+func (rule stringRegexSubrule) Apply(call *ast.CallExpr) {
+	if len(call.Args) <= rule.Scope.Argument {
+		// TODO: should cases where calls are incompatible with scope cause failures?
+		return
+	}
+
+	arg := call.Args[rule.Scope.Argument]
+	var lit *ast.BasicLit
+	if len(rule.Scope.Field) > 0 {
```

Comment:
we could encapsulate this logic into a method that retrieves the `*ast.BasicLit` corresponding to the `rule.Scope.Field`

Reply:
I've tried to work this out, but it just results in more complexity stemming from the fact that we now need to return both the `*ast.BasicLit` and the success of parsing. By keeping this logic in `Apply`, we can simply return early if parsing fails, as opposed to needing to communicate that need upwards.